File: src/query.js
/**
 * Transaction and Output Query Filters
 *
 * Provides flexible filtering for transactions and outputs,
 * similar to monero-ts but adapted for Salvium's transaction types.
 *
 * @module query
 */

import { TX_TYPE } from './wallet.js';

// ============================================================================
// OUTPUT QUERY
// ============================================================================

/**
 * Query filter for wallet outputs (UTXOs)
 */
export class OutputQuery {
  constructor(config = {}) {
    // Amount filters
    this.minAmount = config.minAmount ?? null;
    this.maxAmount = config.maxAmount ?? null;
    this.amount = config.amount ?? null;  // Exact match

    // Status filters
    this.isSpent = config.isSpent ?? null;
    this.isLocked = config.isLocked ?? null;
    this.isFrozen = config.isFrozen ?? null;

    // Location filters
    this.accountIndex = config.accountIndex ?? null;
    this.subaddressIndex = config.subaddressIndex ?? null;
    this.subaddressIndices = config.subaddressIndices ?? null;  // Array of { major, minor }

    // Block filters
    this.minHeight = config.minHeight ?? null;
    this.maxHeight = config.maxHeight ?? null;

    // Key image filter
    this.keyImage = config.keyImage ?? null;
    this.keyImages = config.keyImages ?? null;  // Array

    // Asset type filter (Salvium-specific)
    this.assetType = config.assetType ?? null;

    // Transaction type filter (Salvium-specific)
    this.txType = config.txType ?? null;
    this.txTypes = config.txTypes ?? null;  // Array of TX_TYPE values

    // Transaction hash filter
    this.txHash = config.txHash ?? null;
    this.txHashes = config.txHashes ?? null;  // Array
  }

  /**
   * Test if an output matches this query
   * @param {Object} output - Output to test
   * @param {number} currentHeight - Current blockchain height (for lock status)
   * @param {number} unlockBlocks - Blocks required for unlock (default: 10)
   * @returns {boolean}
   */
  matches(output, currentHeight = 0, unlockBlocks = 10) {
    // Amount filters
    if (this.amount !== null && BigInt(output.amount) !== BigInt(this.amount)) {
      return false;
    }
    if (this.minAmount !== null && BigInt(output.amount) < BigInt(this.minAmount)) {
      return false;
    }
    if (this.maxAmount !== null && BigInt(output.amount) > BigInt(this.maxAmount)) {
      return false;
    }

    // Spent status
    if (this.isSpent !== null && output.isSpent !== this.isSpent) {
      return false;
    }

    // Locked status (computed)
    if (this.isLocked !== null) {
      const isLocked = output.blockHeight &&
        (currentHeight - output.blockHeight) < unlockBlocks;
      if (isLocked !== this.isLocked) {
        return false;
      }
    }

    // Frozen status
    if (this.isFrozen !== null && output.isFrozen !== this.isFrozen) {
      return false;
    }

    // Account filter
    if (this.accountIndex !== null) {
      const major = output.subaddressIndex?.major ?? 0;
      if (major !== this.accountIndex) {
        return false;
      }
    }

    // Subaddress filter
    if (this.subaddressIndex !== null) {
      const minor = output.subaddressIndex?.minor ?? 0;
      if (minor !== this.subaddressIndex) {
        return false;
      }
    }

    // Multiple subaddress filter
    if (this.subaddressIndices !== null) {
      const major = output.subaddressIndex?.major ?? 0;
      const minor = output.subaddressIndex?.minor ?? 0;
      const found = this.subaddressIndices.some(
        idx => idx.major === major && idx.minor === minor
      );
      if (!found) return false;
    }

    // Block height filters
    if (this.minHeight !== null && output.blockHeight < this.minHeight) {
      return false;
    }
    if (this.maxHeight !== null && output.blockHeight > this.maxHeight) {
      return false;
    }

    // Key image filter
    if (this.keyImage !== null) {
      const kiHex = typeof output.keyImage === 'string'
        ? output.keyImage
        : (output.keyImage ? Array.from(output.keyImage).map(b => b.toString(16).padStart(2, '0')).join('') : null);
      if (kiHex !== this.keyImage) {
        return false;
      }
    }

    // Multiple key images filter
    if (this.keyImages !== null) {
      const kiHex = typeof output.keyImage === 'string'
        ? output.keyImage
        : (output.keyImage ? Array.from(output.keyImage).map(b => b.toString(16).padStart(2, '0')).join('') : null);
      if (!kiHex || !this.keyImages.includes(kiHex)) {
        return false;
      }
    }

    // Asset type filter (Salvium)
    if (this.assetType !== null && output.assetType !== this.assetType) {
      return false;
    }

    // Transaction type filter (Salvium)
    if (this.txType !== null && output.txType !== this.txType) {
      return false;
    }
    if (this.txTypes !== null && !this.txTypes.includes(output.txType)) {
      return false;
    }

    // Transaction hash filter
    if (this.txHash !== null && output.txHash !== this.txHash) {
      return false;
    }
    if (this.txHashes !== null && !this.txHashes.includes(output.txHash)) {
      return false;
    }

    return true;
  }

  /**
   * Filter an array of outputs
   * @param {Array<Object>} outputs - Outputs to filter
   * @param {number} currentHeight - Current blockchain height
   * @returns {Array<Object>} Matching outputs
   */
  filter(outputs, currentHeight = 0) {
    return outputs.filter(o => this.matches(o, currentHeight));
  }

  /**
   * Create a copy with additional filters
   * @param {Object} config - Additional filters
   * @returns {OutputQuery}
   */
  with(config) {
    return new OutputQuery({ ...this, ...config });
  }
}

// ============================================================================
// TRANSACTION QUERY
// ============================================================================

/**
 * Query filter for wallet transactions
 */
export class TxQuery {
  constructor(config = {}) {
    // Direction filters
    this.isIncoming = config.isIncoming ?? null;
    this.isOutgoing = config.isOutgoing ?? null;

    // Status filters
    this.isConfirmed = config.isConfirmed ?? null;
    this.inTxPool = config.inTxPool ?? null;
    this.isFailed = config.isFailed ?? null;
    this.isLocked = config.isLocked ?? null;

    // Hash filters
    this.hash = config.hash ?? null;
    this.hashes = config.hashes ?? null;  // Array

    // Block filters
    this.minHeight = config.minHeight ?? null;
    this.maxHeight = config.maxHeight ?? null;
    this.height = config.height ?? null;  // Exact match

    // Amount filters
    this.minAmount = config.minAmount ?? null;
    this.maxAmount = config.maxAmount ?? null;

    // Account filters
    this.accountIndex = config.accountIndex ?? null;
    this.subaddressIndex = config.subaddressIndex ?? null;
    this.subaddressIndices = config.subaddressIndices ?? null;

    // Payment ID filter
    this.hasPaymentId = config.hasPaymentId ?? null;
    this.paymentId = config.paymentId ?? null;
    this.paymentIds = config.paymentIds ?? null;

    // Transaction type filter (Salvium-specific)
    this.txType = config.txType ?? null;
    this.txTypes = config.txTypes ?? null;

    // Asset type filter (Salvium-specific)
    this.assetType = config.assetType ?? null;

    // Include outputs in result
    this.includeOutputs = config.includeOutputs ?? false;
  }

  /**
   * Test if a transaction matches this query
   * @param {Object} tx - Transaction to test
   * @param {number} currentHeight - Current blockchain height
   * @returns {boolean}
   */
  matches(tx, currentHeight = 0) {
    // Direction filters
    if (this.isIncoming !== null && tx.isIncoming !== this.isIncoming) {
      return false;
    }
    if (this.isOutgoing !== null && tx.isOutgoing !== this.isOutgoing) {
      return false;
    }

    // Confirmed status
    if (this.isConfirmed !== null) {
      const isConfirmed = tx.blockHeight !== null && tx.blockHeight !== undefined;
      if (isConfirmed !== this.isConfirmed) {
        return false;
      }
    }

    // In mempool
    if (this.inTxPool !== null && tx.inTxPool !== this.inTxPool) {
      return false;
    }

    // Failed status
    if (this.isFailed !== null && tx.isFailed !== this.isFailed) {
      return false;
    }

    // Locked status
    if (this.isLocked !== null) {
      const unlockBlocks = 10;
      const isLocked = tx.blockHeight &&
        (currentHeight - tx.blockHeight) < unlockBlocks;
      if (isLocked !== this.isLocked) {
        return false;
      }
    }

    // Hash filters
    if (this.hash !== null && tx.hash !== this.hash && tx.txHash !== this.hash) {
      return false;
    }
    if (this.hashes !== null) {
      const txHash = tx.hash || tx.txHash;
      if (!this.hashes.includes(txHash)) {
        return false;
      }
    }

    // Height filters
    if (this.height !== null && tx.blockHeight !== this.height) {
      return false;
    }
    if (this.minHeight !== null && (tx.blockHeight === null || tx.blockHeight < this.minHeight)) {
      return false;
    }
    if (this.maxHeight !== null && (tx.blockHeight === null || tx.blockHeight > this.maxHeight)) {
      return false;
    }

    // Amount filters (total amount)
    const amount = tx.amount ?? tx.totalAmount ?? 0n;
    if (this.minAmount !== null && BigInt(amount) < BigInt(this.minAmount)) {
      return false;
    }
    if (this.maxAmount !== null && BigInt(amount) > BigInt(this.maxAmount)) {
      return false;
    }

    // Account filter
    if (this.accountIndex !== null) {
      // Check if any transfer involves this account
      const hasAccount = (tx.transfers || []).some(t =>
        t.accountIndex === this.accountIndex
      ) || (tx.outputs || []).some(o =>
        o.subaddressIndex?.major === this.accountIndex
      );
      if (!hasAccount) return false;
    }

    // Payment ID filters
    if (this.hasPaymentId !== null) {
      const hasId = tx.paymentId !== null && tx.paymentId !== undefined;
      if (hasId !== this.hasPaymentId) {
        return false;
      }
    }
    if (this.paymentId !== null && tx.paymentId !== this.paymentId) {
      return false;
    }
    if (this.paymentIds !== null && !this.paymentIds.includes(tx.paymentId)) {
      return false;
    }

    // Transaction type filter (Salvium)
    if (this.txType !== null && tx.txType !== this.txType) {
      return false;
    }
    if (this.txTypes !== null && !this.txTypes.includes(tx.txType)) {
      return false;
    }

    // Asset type filter (Salvium)
    if (this.assetType !== null) {
      const hasAsset = (tx.assetType === this.assetType) ||
        (tx.outputs || []).some(o => o.assetType === this.assetType);
      if (!hasAsset) return false;
    }

    return true;
  }

  /**
   * Filter an array of transactions
   * @param {Array<Object>} txs - Transactions to filter
   * @param {number} currentHeight - Current blockchain height
   * @returns {Array<Object>} Matching transactions
   */
  filter(txs, currentHeight = 0) {
    return txs.filter(tx => this.matches(tx, currentHeight));
  }

  /**
   * Create a copy with additional filters
   * @param {Object} config - Additional filters
   * @returns {TxQuery}
   */
  with(config) {
    return new TxQuery({ ...this, ...config });
  }
}

// ============================================================================
// TRANSFER QUERY
// ============================================================================

/**
 * Query filter for transfers (individual movements of funds)
 */
export class TransferQuery {
  constructor(config = {}) {
    // Direction
    this.isIncoming = config.isIncoming ?? null;
    this.isOutgoing = config.isOutgoing ?? null;

    // Amount filters
    this.minAmount = config.minAmount ?? null;
    this.maxAmount = config.maxAmount ?? null;
    this.amount = config.amount ?? null;

    // Address filters
    this.address = config.address ?? null;
    this.addresses = config.addresses ?? null;

    // Account filters
    this.accountIndex = config.accountIndex ?? null;
    this.subaddressIndex = config.subaddressIndex ?? null;
    this.subaddressIndices = config.subaddressIndices ?? null;

    // Asset type (Salvium)
    this.assetType = config.assetType ?? null;

    // Transaction type (Salvium)
    this.txType = config.txType ?? null;
    this.txTypes = config.txTypes ?? null;
  }

  /**
   * Test if a transfer matches this query
   * @param {Object} transfer - Transfer to test
   * @returns {boolean}
   */
  matches(transfer) {
    // Direction
    if (this.isIncoming !== null && transfer.isIncoming !== this.isIncoming) {
      return false;
    }
    if (this.isOutgoing !== null && transfer.isOutgoing !== this.isOutgoing) {
      return false;
    }

    // Amount filters
    if (this.amount !== null && BigInt(transfer.amount) !== BigInt(this.amount)) {
      return false;
    }
    if (this.minAmount !== null && BigInt(transfer.amount) < BigInt(this.minAmount)) {
      return false;
    }
    if (this.maxAmount !== null && BigInt(transfer.amount) > BigInt(this.maxAmount)) {
      return false;
    }

    // Address filters
    if (this.address !== null && transfer.address !== this.address) {
      return false;
    }
    if (this.addresses !== null && !this.addresses.includes(transfer.address)) {
      return false;
    }

    // Account filter
    if (this.accountIndex !== null && transfer.accountIndex !== this.accountIndex) {
      return false;
    }

    // Subaddress filter
    if (this.subaddressIndex !== null && transfer.subaddressIndex !== this.subaddressIndex) {
      return false;
    }

    // Asset type (Salvium)
    if (this.assetType !== null && transfer.assetType !== this.assetType) {
      return false;
    }

    // Transaction type (Salvium)
    if (this.txType !== null && transfer.txType !== this.txType) {
      return false;
    }
    if (this.txTypes !== null && !this.txTypes.includes(transfer.txType)) {
      return false;
    }

    return true;
  }

  /**
   * Filter an array of transfers
   * @param {Array<Object>} transfers - Transfers to filter
   * @returns {Array<Object>} Matching transfers
   */
  filter(transfers) {
    return transfers.filter(t => this.matches(t));
  }
}

// ============================================================================
// CONVENIENCE CONSTRUCTORS
// ============================================================================

/**
 * Create an OutputQuery with given config
 * @param {Object} config - Query config
 * @returns {OutputQuery}
 */
export function createOutputQuery(config = {}) {
  return new OutputQuery(config);
}

/**
 * Create a TxQuery with given config
 * @param {Object} config - Query config
 * @returns {TxQuery}
 */
export function createTxQuery(config = {}) {
  return new TxQuery(config);
}

/**
 * Create a TransferQuery with given config
 * @param {Object} config - Query config
 * @returns {TransferQuery}
 */
export function createTransferQuery(config = {}) {
  return new TransferQuery(config);
}

/**
 * Create an output query for unspent outputs
 * @param {Object} config - Additional config
 * @returns {OutputQuery}
 */
export function unspentOutputs(config = {}) {
  return new OutputQuery({ ...config, isSpent: false });
}

/**
 * Create an output query for spent outputs
 * @param {Object} config - Additional config
 * @returns {OutputQuery}
 */
export function spentOutputs(config = {}) {
  return new OutputQuery({ ...config, isSpent: true });
}

/**
 * Create an output query for locked outputs
 * @param {Object} config - Additional config
 * @returns {OutputQuery}
 */
export function lockedOutputs(config = {}) {
  return new OutputQuery({ ...config, isLocked: true, isSpent: false });
}

/**
 * Create an output query for unlocked outputs
 * @param {Object} config - Additional config
 * @returns {OutputQuery}
 */
export function unlockedOutputs(config = {}) {
  return new OutputQuery({ ...config, isLocked: false, isSpent: false });
}

/**
 * Create an output query for staking outputs (Salvium)
 * @param {Object} config - Additional config
 * @returns {OutputQuery}
 */
export function stakingOutputs(config = {}) {
  return new OutputQuery({
    ...config,
    txType: TX_TYPE.STAKE,
    isSpent: false
  });
}

/**
 * Create an output query for yield payouts (Salvium)
 * @param {Object} config - Additional config
 * @returns {OutputQuery}
 */
export function yieldOutputs(config = {}) {
  return new OutputQuery({
    ...config,
    txType: TX_TYPE.PROTOCOL,
    isSpent: false
  });
}

/**
 * Create a transaction query for incoming transactions
 * @param {Object} config - Additional config
 * @returns {TxQuery}
 */
export function incomingTxs(config = {}) {
  return new TxQuery({ ...config, isIncoming: true });
}

/**
 * Create a transaction query for outgoing transactions
 * @param {Object} config - Additional config
 * @returns {TxQuery}
 */
export function outgoingTxs(config = {}) {
  return new TxQuery({ ...config, isOutgoing: true });
}

/**
 * Create a transaction query for pending (unconfirmed) transactions
 * @param {Object} config - Additional config
 * @returns {TxQuery}
 */
export function pendingTxs(config = {}) {
  return new TxQuery({ ...config, isConfirmed: false, inTxPool: true });
}

/**
 * Create a transaction query for confirmed transactions
 * @param {Object} config - Additional config
 * @returns {TxQuery}
 */
export function confirmedTxs(config = {}) {
  return new TxQuery({ ...config, isConfirmed: true });
}

/**
 * Create a transaction query for staking transactions (Salvium)
 * @param {Object} config - Additional config
 * @returns {TxQuery}
 */
export function stakingTxs(config = {}) {
  return new TxQuery({ ...config, txType: TX_TYPE.STAKE });
}

/**
 * Create a transaction query for yield/protocol transactions (Salvium)
 * @param {Object} config - Additional config
 * @returns {TxQuery}
 */
export function yieldTxs(config = {}) {
  return new TxQuery({ ...config, txType: TX_TYPE.PROTOCOL });
}

export default {
  OutputQuery,
  TxQuery,
  TransferQuery,
  createOutputQuery,
  createTxQuery,
  createTransferQuery,
  unspentOutputs,
  spentOutputs,
  lockedOutputs,
  unlockedOutputs,
  stakingOutputs,
  yieldOutputs,
  incomingTxs,
  outgoingTxs,
  pendingTxs,
  confirmedTxs,
  stakingTxs,
  yieldTxs
};

File: crates/salvium-wallet/src/query.rs
//! Query presets and factory functions for wallet output/transaction filtering.
//!
//! Provides commonly-used query configurations for outputs and transactions.

use salvium_crypto::storage::{OutputQuery, TxQuery};

// ─── Output Query Presets ───────────────────────────────────────────────────

/// Query for unspent, unfrozen outputs.
pub fn unspent_outputs() -> OutputQuery {
    OutputQuery {
        is_spent: Some(false),
        is_frozen: Some(false),
        asset_type: None,
        tx_type: None,
        account_index: None,
        subaddress_index: None,
        min_amount: None,
        max_amount: None,
    }
}

/// Query for spent outputs.
pub fn spent_outputs() -> OutputQuery {
    OutputQuery {
        is_spent: Some(true),
        is_frozen: None,
        asset_type: None,
        tx_type: None,
        account_index: None,
        subaddress_index: None,
        min_amount: None,
        max_amount: None,
    }
}

/// Query for staking outputs (tx_type = 6).
pub fn staking_outputs() -> OutputQuery {
    OutputQuery {
        is_spent: Some(false),
        is_frozen: None,
        asset_type: None,
        tx_type: Some(6),
        account_index: None,
        subaddress_index: None,
        min_amount: None,
        max_amount: None,
    }
}

/// Query for yield/return outputs (tx_type = 7).
pub fn yield_outputs() -> OutputQuery {
    OutputQuery {
        is_spent: Some(false),
        is_frozen: None,
        asset_type: None,
        tx_type: Some(7),
        account_index: None,
        subaddress_index: None,
        min_amount: None,
        max_amount: None,
    }
}

/// Query for outputs of a specific asset type.
pub fn outputs_by_asset(asset_type: &str) -> OutputQuery {
    OutputQuery {
        is_spent: None,
        is_frozen: None,
        asset_type: Some(asset_type.to_string()),
        tx_type: None,
        account_index: None,
        subaddress_index: None,
        min_amount: None,
        max_amount: None,
    }
}

/// Query for outputs in an amount range.
pub fn outputs_in_range(min: u64, max: u64) -> OutputQuery {
    OutputQuery {
        is_spent: None,
        is_frozen: None,
        asset_type: None,
        tx_type: None,
        account_index: None,
        subaddress_index: None,
        min_amount: Some(min.to_string()),
        max_amount: Some(max.to_string()),
    }
}

/// Query for outputs by account.
pub fn outputs_by_account(account_index: i64) -> OutputQuery {
    OutputQuery {
        is_spent: Some(false),
        is_frozen: None,
        asset_type: None,
        tx_type: None,
        account_index: Some(account_index),
        subaddress_index: None,
        min_amount: None,
        max_amount: None,
    }
}

// ─── Transaction Query Presets ──────────────────────────────────────────────

/// Query for incoming transactions.
pub fn incoming_txs() -> TxQuery {
    TxQuery {
        is_incoming: Some(true),
        is_outgoing: None,
        is_confirmed: None,
        in_pool: None,
        tx_type: None,
        min_height: None,
        max_height: None,
        tx_hash: None,
    }
}

/// Query for outgoing transactions.
pub fn outgoing_txs() -> TxQuery {
    TxQuery {
        is_incoming: None,
        is_outgoing: Some(true),
        is_confirmed: None,
        in_pool: None,
        tx_type: None,
        min_height: None,
        max_height: None,
        tx_hash: None,
    }
}

/// Query for pending (in-pool) transactions.
pub fn pending_txs() -> TxQuery {
    TxQuery {
        is_incoming: None,
        is_outgoing: None,
        is_confirmed: None,
        in_pool: Some(true),
        tx_type: None,
        min_height: None,
        max_height: None,
        tx_hash: None,
    }
}

/// Query for confirmed transactions.
pub fn confirmed_txs() -> TxQuery {
    TxQuery {
        is_incoming: None,
        is_outgoing: None,
        is_confirmed: Some(true),
        in_pool: None,
        tx_type: None,
        min_height: None,
        max_height: None,
        tx_hash: None,
    }
}

/// Query for staking transactions (tx_type = 6).
pub fn staking_txs() -> TxQuery {
    TxQuery {
        is_incoming: None,
        is_outgoing: None,
        is_confirmed: None,
        in_pool: None,
        tx_type: Some(6),
        min_height: None,
        max_height: None,
        tx_hash: None,
    }
}

/// Query for yield/return transactions (tx_type = 7).
pub fn yield_txs() -> TxQuery {
    TxQuery {
        is_incoming: None,
        is_outgoing: None,
        is_confirmed: None,
        in_pool: None,
        tx_type: Some(7),
        min_height: None,
        max_height: None,
        tx_hash: None,
    }
}

/// Query for transactions in a height range.
pub fn txs_in_height_range(min: i64, max: i64) -> TxQuery {
    TxQuery {
        is_incoming: None,
        is_outgoing: None,
        is_confirmed: None,
        in_pool: None,
        tx_type: None,
        min_height: Some(min),
        max_height: Some(max),
        tx_hash: None,
    }
}

/// Query for a specific transaction by hash.
pub fn tx_by_hash(hash: &str) -> TxQuery {
    TxQuery {
        is_incoming: None,
        is_outgoing: None,
        is_confirmed: None,
        in_pool: None,
        tx_type: None,
        min_height: None,
        max_height: None,
        tx_hash: Some(hash.to_string()),
    }
}

// ─── Tests ──────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // ── Output Query Preset Tests ───────────────────────────────────────

    #[test]
    fn test_unspent_outputs_preset() {
        let q = unspent_outputs();
        assert_eq!(q.is_spent, Some(false));
        assert_eq!(q.is_frozen, Some(false));
        assert!(q.asset_type.is_none());
        assert!(q.tx_type.is_none());
        assert!(q.account_index.is_none());
        assert!(q.subaddress_index.is_none());
        assert!(q.min_amount.is_none());
        assert!(q.max_amount.is_none());
    }

    #[test]
    fn test_spent_outputs_preset() {
        let q = spent_outputs();
        assert_eq!(q.is_spent, Some(true));
        assert!(q.is_frozen.is_none());
        assert!(q.asset_type.is_none());
        assert!(q.tx_type.is_none());
        assert!(q.account_index.is_none());
        assert!(q.subaddress_index.is_none());
        assert!(q.min_amount.is_none());
        assert!(q.max_amount.is_none());
    }

    #[test]
    fn test_staking_outputs_preset() {
        let q = staking_outputs();
        assert_eq!(q.is_spent, Some(false));
        assert_eq!(q.tx_type, Some(6));
        assert!(q.is_frozen.is_none());
        assert!(q.asset_type.is_none());
    }

    #[test]
    fn test_yield_outputs_preset() {
        let q = yield_outputs();
        assert_eq!(q.is_spent, Some(false));
        assert_eq!(q.tx_type, Some(7));
        assert!(q.is_frozen.is_none());
        assert!(q.asset_type.is_none());
    }

    #[test]
    fn test_outputs_by_asset() {
        let q = outputs_by_asset("SAL");
        assert_eq!(q.asset_type, Some("SAL".to_string()));
        assert!(q.is_spent.is_none());
        assert!(q.is_frozen.is_none());
        assert!(q.tx_type.is_none());

        let q2 = outputs_by_asset("USD");
        assert_eq!(q2.asset_type, Some("USD".to_string()));
    }

    #[test]
    fn test_outputs_in_range() {
        let q = outputs_in_range(1000, 50000);
        assert_eq!(q.min_amount, Some("1000".to_string()));
        assert_eq!(q.max_amount, Some("50000".to_string()));
        assert!(q.is_spent.is_none());
        assert!(q.asset_type.is_none());
        assert!(q.tx_type.is_none());
    }

    #[test]
    fn test_outputs_by_account() {
        let q = outputs_by_account(2);
        assert_eq!(q.account_index, Some(2));
        assert_eq!(q.is_spent, Some(false));
        assert!(q.is_frozen.is_none());
        assert!(q.subaddress_index.is_none());
    }

    // ── Transaction Query Preset Tests ──────────────────────────────────

    #[test]
    fn test_incoming_txs_preset() {
        let q = incoming_txs();
        assert_eq!(q.is_incoming, Some(true));
        assert!(q.is_outgoing.is_none());
        assert!(q.is_confirmed.is_none());
        assert!(q.in_pool.is_none());
        assert!(q.tx_type.is_none());
        assert!(q.min_height.is_none());
        assert!(q.max_height.is_none());
        assert!(q.tx_hash.is_none());
    }

    #[test]
    fn test_outgoing_txs_preset() {
        let q = outgoing_txs();
        assert_eq!(q.is_outgoing, Some(true));
        assert!(q.is_incoming.is_none());
        assert!(q.is_confirmed.is_none());
        assert!(q.in_pool.is_none());
    }

    #[test]
    fn test_pending_txs_preset() {
        let q = pending_txs();
        assert_eq!(q.in_pool, Some(true));
        assert!(q.is_incoming.is_none());
        assert!(q.is_outgoing.is_none());
        assert!(q.is_confirmed.is_none());
        assert!(q.tx_type.is_none());
    }

    #[test]
    fn test_confirmed_txs_preset() {
        let q = confirmed_txs();
        assert_eq!(q.is_confirmed, Some(true));
        assert!(q.is_incoming.is_none());
        assert!(q.is_outgoing.is_none());
        assert!(q.in_pool.is_none());
    }

    #[test]
    fn test_staking_txs_preset() {
        let q = staking_txs();
        assert_eq!(q.tx_type, Some(6));
        assert!(q.is_incoming.is_none());
        assert!(q.is_outgoing.is_none());
        assert!(q.is_confirmed.is_none());
        assert!(q.in_pool.is_none());
    }

    #[test]
    fn test_yield_txs_preset() {
        let q = yield_txs();
        assert_eq!(q.tx_type, Some(7));
        assert!(q.is_incoming.is_none());
        assert!(q.is_outgoing.is_none());
        assert!(q.is_confirmed.is_none());
        assert!(q.in_pool.is_none());
    }

    #[test]
    fn test_txs_in_height_range() {
        let q = txs_in_height_range(100, 500);
        assert_eq!(q.min_height, Some(100));
        assert_eq!(q.max_height, Some(500));
        assert!(q.is_incoming.is_none());
        assert!(q.is_outgoing.is_none());
        assert!(q.tx_type.is_none());
        assert!(q.tx_hash.is_none());
    }

    #[test]
    fn test_tx_by_hash() {
        let q = tx_by_hash("abc123def456");
        assert_eq!(q.tx_hash, Some("abc123def456".to_string()));
        assert!(q.is_incoming.is_none());
        assert!(q.is_outgoing.is_none());
        assert!(q.is_confirmed.is_none());
        assert!(q.in_pool.is_none());
        assert!(q.tx_type.is_none());
        assert!(q.min_height.is_none());
        assert!(q.max_height.is_none());
    }
}
